src/emqx_hooks.erl

```diff
@@ -16,142 +16,160 @@
         ok             -> run_(Callbacks, Args, Acc);
         {ok, NewAcc}   -> run_(Callbacks, Args, NewAcc);
         stop           -> {stop, Acc};
         {stop, NewAcc} -> {stop, NewAcc};
         _Any           -> run_(Callbacks, Args, Acc)
     end;
-
 run_([], _Args, Acc) ->
     {ok, Acc}.
 
--spec(lookup(atom()) -> [#callback{}]).
+filtered(undefined, _Args) ->
+    false;
+filtered(Filter, Args) ->
+    execute(Filter, Args).
+
+execute(Action, Args) when is_function(Action) ->
+    erlang:apply(Action, Args);
+execute({Fun, InitArgs}, Args) when is_function(Fun) ->
+    erlang:apply(Fun, Args ++ InitArgs);
+execute({M, F, A}, Args) ->
+    erlang:apply(M, F, Args ++ A).
+
+%% @doc Lookup callbacks.
+-spec(lookup(hookpoint()) -> [#callback{}]).
 lookup(HookPoint) ->
     case ets:lookup(?TAB, HookPoint) of
-        [#hook{callbacks = Callbacks}] -> Callbacks;
+        [#hook{callbacks = Callbacks}] ->
+            Callbacks;
         [] -> []
     end.
 
-%%--------------------------------------------------------------------
+%%------------------------------------------------------------------------------
 %% gen_server callbacks
-%%--------------------------------------------------------------------
+%%------------------------------------------------------------------------------
 
 init([]) ->
-    _ = emqx_tables:new(?TAB, [set, protected, {keypos, #hook.name},
-                               {read_concurrency, true}]),
-    {ok, #state{}}.
-
-handle_call({add, HookPoint, {Tag, Function}, InitArgs, Priority}, _From, State) ->
-    Callback = #callback{tag = Tag, function = Function,
-                         init_args = InitArgs, priority = Priority},
-    {reply,
-     case ets:lookup(?TAB, HookPoint) of
-         [#hook{callbacks = Callbacks}] ->
-             case contain_(Tag, Function, Callbacks) of
-                 false ->
-                     insert_hook_(HookPoint, add_callback_(Callback, Callbacks));
-                 true  ->
-                     {error, already_hooked}
-             end;
-         [] ->
-             insert_hook_(HookPoint, [Callback])
-     end, State};
-
-handle_call({delete, HookPoint, {Tag, Function}}, _From, State) ->
-    {reply,
-     case ets:lookup(?TAB, HookPoint) of
-         [#hook{callbacks = Callbacks}] ->
-             case contain_(Tag, Function, Callbacks) of
-                 true  ->
-                     insert_hook_(HookPoint, del_callback_(Tag, Function, Callbacks));
-                 false ->
-                     {error, not_found}
-             end;
-         [] ->
-             {error, not_found}
-     end, State};
+    _ = emqx_tables:new(?TAB, [{keypos, #hook.name}, {read_concurrency, true}]),
+    {ok, #{}}.
+
+handle_call({add, HookPoint, Callback = #callback{action = Action}}, _From, State) ->
+    Reply = case lists:keymember(Action, 2, Callbacks = lookup(HookPoint)) of
+                true ->
+                    {error, already_exists};
```

Comment:
what is the caller supposed to do on this error?
if they are mostly to treat this as an `ok`, perhaps change the API name from `add` to `ensure_present`